### ISSUE TITLE
Provide parameters to create service backed by operators

### ISSUE THREAD

**Issue by dharmit** [openshift/odo#2785]
/kind user-story

## User Story
As a user I want to be able to pass parameters to service creation command (`odo service create`) when creating an operator backed service so that I can spin up the service quickly and exactly as I need the various configurations to be setup.

## Acceptance Criteria 
- [ ] `odo service create <operator-name> --crd <crd-name> -p param1=value1 -p param2=value2` should spin up a service based on `crd-name` from the operator `operator-name` and configure the mentioned parameters to the specified values.

## Links 
- Feature Request: #2613

/kind user-story

**Comment by girishramnani** [comment]
so figured out from this https://github.com/openshift/console/pull/2580 that the `openapi/v2` exposes the `openapischema` as well. 
But this only happens on openshift, but also k8s allows getting CRDs without any RBAC problem. so we have to use openapi swagger for openshift and CRD def for k8s.
the swagger - https://raw.githubusercontent.com/openshift/origin/master/api/swagger-spec/openshift-openapi-spec.json

So the way it seems to happen is that openapi/v2 gets updated based on the latest operators being added.

**Comment by girishramnani** [comment]
reason we need different approach for k8s because CRD defs are not published to `openapi/v2` schema https://github.com/openshift/console/pull/2580/files#r320472178

**Comment by dharmit** [comment]
> the swagger - https://raw.githubusercontent.com/openshift/origin/master/api/swagger-spec/openshift-openapi-spec.json

How do I find the fields for the CRs provided by each CSV here? I tried searching for etcd but there doesn't seem to be any useful info here. Maybe I'm not looking at it correctly.

The way I have been looking at this is that Web Console gives us an option of "Edit Form" when we try to create a service from a CR. @kadel pointed to the same as well. The info on that page comes from the CSV's YAML which is available to us as regular user.

For example, if you would like to create EtcdCluster service, you'd see this when you click on "Edit Form":
![Screenshot_2020-07-16 Create EtcdCluster · Red Hat OpenShift Container Platform](https://user-images.githubusercontent.com/713521/87662539-fbf2f880-c77f-11ea-9720-a7617fc0d6fb.png)

The fields we see in above image are mentioned in `specDescriptors` field in the YAML of the CSV:
```yaml
customresourcedefinitions:
    owned:
    - description: Represents a cluster of etcd nodes.
      displayName: etcd Cluster
      kind: EtcdCluster
      name: etcdclusters.etcd.database.coreos.com
      resources:
      - kind: Service
        name: ""
        version: v1
      - kind: Pod
        name: ""
        version: v1
      specDescriptors:
      - description: The desired number of member Pods for the etcd cluster.
        displayName: Size
        path: size
        x-descriptors:
        - urn:alm:descriptor:com.tectonic.ui:podCount
      - description: Limits describes the minimum/maximum amount of compute resources
          required/allowed
        displayName: Resource Requirements
        path: pod.resources
        x-descriptors:
        - urn:alm:descriptor:com.tectonic.ui:resourceRequirements
```

The field used by Web Console is (probably) picked from the `x-descriptors` field. That helps give a UX we can't provide through CLI. OTOH, if we let the user set these values, we need to validate them or trust that the user entered sane values.

Besides that, the Web Console has a recommendation at the right about using editing the YAML for better control of the final object that would get created.

**Comment by dharmit** [comment]
> The field used by Web Console is (probably) picked from the `x-descriptors` field. That helps give a UX we can't provide through CLI.

By "field" I mean the UI that we see on Web Console.



> Besides that, the Web Console has a recommendation at the right about using editing the YAML for better control of the final object that would get created.

While mentioning this, I was thinking about interactive mode for service create. We need to study a few CSVs and the CRs they provide to be able to take either this or #2799 further.

**Comment by dharmit** [comment]
/remove-lifecycle stale

@girishramnani you recently mentioned that you found something relevant to this. Can you share on the issue?

**Comment by girishramnani** [comment]
Actually this issue needs to be broken down into one for kubernetes and one for openshift due this comment -

> reason we need different approach for k8s because CRD defs are not published to `openapi/v2` schema https://github.com/openshift/console/pull/2580/files#r320472178

I would be creating an interface `OperatorMetadataProvider` that would be implemented for both kubernetes and openshift and a builder could check the presence of the `api` and return the desired Provider.

**Comment by girishramnani** [comment]
working on https://github.com/openshift/odo/issues/2615 first as it seems to pre-requisite to this issue's AC

**Comment by girishramnani** [comment]
## Scope for this sprint

TLDR - Allow user to use the information from `odo catalog describe service <operator>` to provide parameters when creating ( adding to devfile ) an operand `odo service create <operator> <name> -p <param>=<value>`.

### Implementation details

So taking an example of `PgCluster`
```
odo catalog describe service postgresoperator.v4.5.0/Pgcluster      
Kind: Pgcluster
Description: Represents a Postgres primary cluster member
Parameters:
- Path: ccpimage
  DisplayName: PostgreSQL Image
  Description: The Crunchy PostgreSQL image to use. Possible values are "crunchy-postgres-ha" and "crunchy-postgres-gis-ha"
- Path: ccpimagetag
  DisplayName: PostgreSQL Image Tag
  Description: The tag of the PostgreSQL image to use. Example is "ubi7-12.4-4.5.0"
- Path: clustername
  DisplayName: PostgreSQL Cluster name
  Description: The name that is assigned to this specific PostgreSQL cluster
- Path: database
  DisplayName: Initial PostgreSQL database name
  Description: The name of the initial database to be created inside of the PostgreSQL cluster, e.g. "hippo"
- Path: exporterport
  DisplayName: PostgreSQL Monitor Port
  Description: The port to use for the PostgreSQL metrics exporter used for cluster monitoring, e.g. "9187"
- Path: name
  DisplayName: PostgreSQL CRD name
  Description: The name of the CRD entry. Should match the PostgreSQL Cluster name
- Path: pgbadgerport
  DisplayName: pgBadger Port
  Description: The port to use for the pgBadger PostgreSQL query analysis service, e.g. "10000"
- Path: port
  DisplayName: PostgreSQL Port
  Description: The port to use for the PostgreSQL cluster, e.g. "5432"
- Path: rootsecretname
  DisplayName: PostgreSQL superuser credentials
  Description: The name of the Secret that contains the PostgreSQL superuser credentials
- Path: primarysecretname
  DisplayName: PostgreSQL support service credentials
  Description: The name of the Secret that contains the credentials used for managing cluster instance authentication, e.g. connections for replicas
- Path: usersecretname
  DisplayName: PostgreSQL user credentials
  Description: The name of the Secret that contains the PostgreSQL user credentials for logging into the PostgreSQL cluster
- Path: PrimaryStorage
  DisplayName: PostgreSQL Primary Storage
  Description: Attributes that help set the primary storage of a PostgreSQL cluster
- Path: PrimaryStorage.name
  DisplayName: PostgreSQL Primary Storage Name
  Description: Contains the name of the PostgreSQL cluster to associate with this storage. Should match the Cluster name
- Path: PrimaryStorage.storageclass
  DisplayName: PostgreSQL Primary StorageClass
  Description: Contains the storage class used for the primary PostgreSQL instance of the cluster
- Path: PrimaryStorage.accessmode
  DisplayName: PostgreSQL Primary StorageClass Access Mode
  Description: The access mode for the storage class, e.g. "ReadWriteOnce"
- Path: PrimaryStorage.size
  DisplayName: PostgreSQL Primary Data PVC Size
  Description: The size of the PVC that will store the data for the primary PostgreSQL instance, e.g. "1G"
- Path: status
  DisplayName: Deprecated
  Description: Deprecated
- Path: userlabels
  DisplayName: User defined labels
  Description: A set of labels that help the PostgreSQL Operator manage a PostgreSQL cluster
```

Next step is to allow user to set these values using `odo service create`. 

The Acceptance Criteria is as follows -

- User can provide the `path`'s value for any param from the describe output and set it in `odo service create`
E.g. `odo service create postgresoperator.v4.5.0/Pgcluster -p PrimaryStorage.size=1G -p port=5455`
- We need to validate these values when the user provides them _case insensitively_.
-  if a parameter isn't present in the description of the CR then odo should fail with error `provided param %s is supported by <operator>`
- Every parameter would be considered a string for now, stretch goal would be to detect "resourceRequirements" struct as that is used at many places. 

### Questions

- should we validate when `--from-file` param is used and user provides a file that contains the CRD, should we validate that against the description? Or we assume the user is a power user and they know what they are doing?

### Out of scope

- Autocomplete for params
- interactive mode with enhanced params
- anything related to service catalog

**Comment by girishramnani** [comment]
Planning to use https://github.com/tidwall/sjson to build the json structure based on the input. 
sjson is really powerful and might allow use to leverage more of its syntax in the future. We can link its syntax rules in odo.

**Comment by dharmit** [comment]
>  should we validate when --from-file param is used and user provides a file that contains the CRD, should we validate that against the description? Or we assume the user is a power user and they know what they are doing?

Not in the first pass.

**Comment by girishramnani** [comment]
so I am able to set values in the spec and also store the cr in devfile but I dont think we can go with all string approach
```
reating Kubernetes resources for component erte
 ✗  Failed to start component with name "erte". Error: Failed to create the component: failed to create service(s) associated with the component: Kafka.kafka.strimzi.io "blah" is invalid: [spec.kafka.replicas: Invalid value: "string": spec.kafka.replicas in body must be of type integer: "string", spec.kafka.listeners: Required value, spec.kafka.storage: Required value, spec.zookeeper: Required value]
```

**Comment by girishramnani** [comment]
so it seems there are certain parameters that are required
```
 spec.kafka.listeners: Required value
```

but these dont show in description

```
Kind: Kafka
Version: v1beta2
Description: Represents a Kafka cluster
Parameters:
- Path: kafka.version
  DisplayName: Version
  Description: Kafka version
- Path: kafka.replicas
  DisplayName: Kafka Brokers
  Description: The desired number of Kafka brokers.
- Path: kafka.storage.type
  DisplayName: Kafka storage
  Description: The type of storage used by Kafka brokers
- Path: kafka.resources
  DisplayName: Kafka Resource Requirements
  Description: Limits describes the minimum/maximum amount of compute resources required/allowed
- Path: zookeeper.replicas
  DisplayName: Zookeeper Nodes
  Description: The desired number of Zookeeper nodes.
- Path: zookeeper.storage.type
  DisplayName: Zookeeper storage
  Description: The type of storage used by Zookeeper nodes
- Path: zookeeper.resources
  DisplayName: Zookeeper Resource Requirements
  Description: Limits describes the minimum/maximum amount of compute resources required/allowed
```

We need to re-consider the source of description

**Comment by girishramnani** [comment]
operator `strimzi-cluster-operator.v0.23.0/Kafka `

**Comment by girishramnani** [comment]
Or we only allow certain kinds of operators?

**Comment by girishramnani** [comment]
Or I can parse the alm examples and build a hybrid description.
And then use alm examples as the base layer with default in the first iteration?

**Comment by kadel** [comment]
> so I am able to set values in the spec and also store the cr in devfile but I dont think we can go with all string approach

if user specifies integer we should use integer.
For example. If user sets `-p spec.kafka.replicas 1` it should be set as an integer. If the user specifies `-p spec.kafka.replicas "1"` it should be set as a string.

**Comment by kadel** [comment]
> Also it seems there are certain parameters that are required

Are there other places where we can get that information?


> Or we only allow certain kinds of operators?

How would we decide what kind of operators we allow?

> Or I can parse the alm examples and build a hybrid description.
> And then use alm examples as the base layer with default in the first iteration?

How common is it for operators to have alm examples? If I remember correctly most of the operators don't have that.



But I don't think that this is that big of a problem right now. Yes, the user experience won't be great. But it looks like at least we can show users what are the required parameters after they try to create the operator for the first time. Then they will know all required parameters and they can run `odo service create` command again and specify all that is required.

**Comment by girishramnani** [comment]
> How common is it for operators to have alm examples? If I remember correctly most of the operators don't have that.

I have seen them in many community operators but I am suggesting that we use them when we have them in combination with spec descriptors.

**Comment by kadel** [comment]
> I have seen them in many community operators but I am suggesting that we use them when we have them in combination with spec descriptors.

lets to that in the next iterations as an improvement of this feature.

**Comment by kadel** [comment]
This is not really a blocker for the first iteration of this feature. The experience doesn't have to be ideal from the beginning. We can figure out later how to show the required parameters in advance.

**Comment by kadel** [comment]
> But I don't think that this is that big of a problem right now. Yes, the user experience won't be great. But it looks like at least we can show users what are the required parameters after they try to create the operator for the first time. Then they will know all required parameters and they can run `odo service create` command again and specify all that is required.

I just realized that this would require that the whole CR gets validated during `odo service create` .
Can we do that?
Even if we can do that it needs to be "optional" check. As `odo service create needs to work even if user is not connected to the cluster.

**Comment by girishramnani** [comment]
But odo service create doesn't work without cluster even before this change and I dont think it can because we need to download the `alm` examples from the cluster. also the CSV from the cluster gets downloaded and validated on `odo service create`

Build on odo main branch
```
make install
go: downloading github.com/tidwall/gjson v1.7.3
go install -mod=vendor -ldflags="-X github.com/openshift/odo/pkg/version.GITCOMMIT=bd2100d91" ./cmd/odo/
(base) girish@MacBook-Pro odo % cd te2 
(base) girish@MacBook-Pro te2 % export KUBECONFIG=/bo/sdf
(base) girish@MacBook-Pro te2 % odo service create strimzi-cluster-operator.v0.23.0/Kafka -p kafka.replicas=1 blah
W0522 11:47:34.593126   14507 loader.go:223] Config not found: /bo/sdf
```

**Comment by girishramnani** [comment]
> > so I am able to set values in the spec and also store the cr in devfile but I dont think we can go with all string approach
> 
> if user specifies integer we should use integer.
> For example. If user sets `-p spec.kafka.replicas 1` it should be set as an integer. If the user specifies `-p spec.kafka.replicas "1"` it should be set as a string.

I tried doing this and 
`"1"` is actually gets interpolated by bash and the application directly just gets `1`, the only way I was able to send `"1"` is when I did `spec.kafka.replicas '"1"'` and that is not good ux.
So I was thinking if I try to parse these values like in order of
- try to parse as an integer, if that fails then assume string.

And if a user wants to specify an integer that gets set as a string then they can just do `'"1"'` for such edge cases.

e.g. `spec.kafka.replicas "1"` here 1 would be parsed as an integer as it would do so without error.

**Comment by girishramnani** [comment]
So the current approach is in two parts
- we do the optimistic type checking mentioned in the comment above but with more types, we will check ( and this is in order )
    - try to parse as integer, if it goes through then we assume it as an integer
    - try to parse as a float and if it goes through the we assume the value as float
    - then everything else becomes string.
    - note - booleans are automatically parsed from `true` and `false` from yaml when being applied on cluster so we dont have to do anything with that.

- This is a working theory - we will try to do a "dry run" create in the `odo service create` and confirm if the crd is valid or not on `odo service create` itself. this will show all required fields and errors without doing `odo push`

**Comment by girishramnani** [comment]
found something super interesting. we can fetch the openapiv3 schema of a crd without admin mode.

<cluster>/kubernetes/openapi/v2
But its a 15MB file which holds swagger doc for the while cluster

**Comment by girishramnani** [comment]
this is the whole schema https://gist.github.com/girishramnani/f29949d5cb8c6547102776437e05ac19

**Comment by girishramnani** [comment]
The proposal has been merged https://github.com/openshift/odo/pull/4751
